framework/data_types/ndarray.h

```diff
@@ -446,94 +242,104 @@ class NDArray
-    return *(address);
+    return storage_[compute_index(args...)];
   }
 
   /**
-   * Accesses the specified element with safety checks.
+   * Accesses the specified element for an array with N dimensions with bounds checking.
    *
-   * \param args The indices of the desired element.
-   * \throw std::invalid_argument if the number of arguments are incorrect and
-   * std::out_of_range if one of the dimension-indices are out of range.
-   * \return Read/write reference to the element.
+   * \param args Indices for each dimension.
+   * \return Read reference to the element.
    */
   template <typename... Args>
-  T& at(Args... args) noexcept
+  constexpr const T& at(Args... args) const
   {
-    static_assert(AllIntegral<Args...>::value,
-                  "NDArray::at(): All parameters must be of integral type");
-
-    if (sizeof...(args) != rank_)
-      throw std::invalid_argument("NDArray::at(): Number of arguments " +
-                                  std::to_string(sizeof...(args)) + " not equal to rank " +
-                                  std::to_string(rank_));
-
-    const size_t N = rank_;
     size_t indices[]{static_cast<size_t>(args)...};
-    for (size_t i = 0; i < N; ++i)
+    for (size_t i = 0; i < rank_; ++i)
+    {
       if (indices[i] >= dimensions_[i])
-        throw std::out_of_range("NDArray::at(): Index " + std::to_string(i) + " out of range " +
-                                std::to_string(indices[i]) + " must be <" +
-                                std::to_string(dimensions_[i]));
+        throw std::out_of_range("Index out of bounds.");
+    }
+    return (*this)(args...);
+  }
 
-    T* address = base_ + indices[N - 1];
-    for (size_t i = 0; i < N - 1; ++i)
-      address += strides_[i] * indices[i];
-    return *(address);
+  /// Returns an iterator pointing to the beginning of the array.
+  inline constexpr T* begin() const noexcept { return storage_.get(); }
+
+  /// Returns a constant iterator pointing to the beginning of the array.
+  inline constexpr const T* cbegin() const noexcept { return storage_.get(); }
+
+  /// Returns an iterator pointing to the end of the array.
+  inline constexpr T* end() const noexcept { return storage_.get() + size_; }
+
+  /// Returns a constant iterator pointing to the end of the array.
+  inline constexpr const T* cend() const noexcept { return storage_.get() + size_; }
+
+  /// Returns the number of elements in the array.
+  inline constexpr size_t size() const noexcept { return size_; }
+
+  /// Returns true if the array has no elements
+  inline constexpr bool empty() const noexcept { return size_ == 0; }
+
+  /// Returns a pointer to the underlying array data.
+  inline constexpr T* data() const noexcept { return storage_.get(); }
+
+  /// Returns the rank of the array.
+  inline constexpr size_t rank() const noexcept { return rank_; }
+
+  /// Returns the dimension of the array.
+  inline std::vector<size_t> dimension() const noexcept
```

Comment:
With `size_t` used here, should we have the template parameter `D` in routines above or just use `size_t`?

Reply:
I think just size_t. For this purpose, I don't think it's important to template.